### PR TITLE
[Bridges] use dual of equality in slack bridge (take 2)

### PR DESCRIPTION
**Pull request by blegat**
We can take inspiration from https://github.com/jump-dev/MathOptInterface.jl/blob/08d34003cd3d7500379d9440e92445bc4d9a8fa8/src/Bridges/Constraint/bridges/set_dot_scaling.jl#L64-L75

Closes https://github.com/jump-dev/MathOptInterface.jl/issues/2513

### CODE REVIEW

File: src/Bridges/Constraint/bridges/slack.jl [hunk 1]
```diff
@@ -124,12 +124,20 @@ end
 function MOI.get(
     model::MOI.ModelLike,
     a::Union{MOI.ConstraintDual,MOI.ConstraintDualStart},
-    bridge::_AbstractSlackBridge,
-)
+    bridge::_AbstractSlackBridge{T},
+) whhere {T}
```
Comment (odow) [inline src/Bridges/Constraint/bridges/slack.jl:128]:
```suggestion
) where {T}
```

**Comment by odow** [comment]
https://github.com/jump-dev/MathOptInterface.jl/actions/runs/9531770131

**Comment by odow** [comment]
Still the same problem: https://github.com/jump-dev/MathOptInterface.jl/actions/runs/9531770131/job/26273074388

image

There's a suspicious `1 / sqrt(2)` there, so perhaps we need `1 / scale^2` somewhere?

**Comment by odow** [comment]
https://github.com/jump-dev/MathOptInterface.jl/actions/runs/9539455292

File: src/Bridges/Constraint/bridges/slack.jl [hunk 2]
```diff
@@ -124,12 +124,25 @@ end
...
+    # does not # support `ConstraintDual`. This is the case if the adjoint of
+    # the linear map on which the bridge is based is not invertible, as for
+    # instance with `Variable.ZerosBridge` or
+    # `SumOfSquares.Bridges.Variable.KernelBridge`.
+    dual = MOI.get(model, a, bridge.equality)
+    if dual === nothing
+        return nothing
+    elseif dual isa AbstractVector
+        scale = MOI.Utilities.SetDotScalingVector{T}(bridge.set)
+        return dual ./ scale.^2
```
Comment (odow) [inline src/Bridges/Constraint/bridges/slack.jl:143]:
```suggestion
        return dual ./ scale .^ 2
```

**Comment by odow** [comment]
The COPT failure is unrelated: https://github.com/COPT-Public/COPT.jl/pull/40

**Comment by odow** [comment]
I don't fully understand the math logic, but the code does something like `set_dot(scale * x, scale * y)`, which is more or less equivalent to `sum(scale.^2 .* x .* y)`, so to reverse, we need to divide by `scale.^2`.

**Comment by blegat** [comment]
Maybe we should add a test in `test/Bridges/Constraint/slack.jl`

**Comment by odow** [comment]
I've added a test that triggers the `scale` and tweaked the comments. I also needed to update the `MOI.set` for `::ConstraintDualStart` because it didn't scale things either.

As a more general comment: our tests just check that `get/set` is a round-trip. We can't easily automatically test that the numeric values are correct.

**Review by odow — APPROVED** [review]

**Comment by odow** [comment]
Let's just double check:

https://github.com/jump-dev/MathOptInterface.jl/actions/runs/9556803924

File: src/Bridges/Constraint/bridges/slack.jl [hunk 3]
```diff
@@ -124,26 +124,46 @@ end
...
     value,
-)
-    # As the slack appears `+slack` in `slack_in_set` and `-slack` in equality,
-    # giving `value` to both will cancel it out in the Lagrangian.
-    # Giving `value` to `bridge.equality` will put the function in the
-    # Lagrangian as expected.
+) where {T}
+    # See comments in MOI.get for why we need to rescale, etc.
     MOI.set(model, attr, bridge.slack_in_set, value)
-    MOI.set(model, attr, bridge.equality, value)
```
Comment (blegat) [inline src/Bridges/Constraint/bridges/slack.jl:146]: Ouch, this was a bug, good catch

**Pull request merged**